fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/api/attachment/v1/AttachmentContainer.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.attachment.v1;
+
+import java.util.function.Supplier;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.nbt.NbtCompound;
+
+import net.fabricmc.fabric.impl.attachment.AttachmentContainerImpl;
+
+/**
+ * Maps {@linkplain AttachmentType attachment types} to attached values.
+ */
+@ApiStatus.NonExtendable
+public interface AttachmentContainer {
+	/**
+	 * Create a new attachment container.
+	 */
+	static AttachmentContainer create() {
+		return new AttachmentContainerImpl();
+	}
+
+	/* QUERY METHODS */
+
+	/**
+	 * Return the value associated with the passed attachment type, or {@code null} if it's absent.
+	 */
+	@Nullable
+	<T> T get(AttachmentType<T> attachment);
+
+	/**
+	 * If the passed attachment type is not associated to a value, compute its value using the supplier and enter it into the container.
+	 * Then return the value in this container.
+	 */
+	default <T> T computeIfAbsent(AttachmentType<T> attachment, Supplier<T> supplier) {
+		T value = get(attachment);
+
+		if (value == null) {
+			set(attachment, value = supplier.get());
+		}
+
+		return value;
+	}
+
+	/**
+	 * Associate a value to the passed attachment type, and return the previous value or {@code null} if there was no previous value.
+	 */
+	@Nullable
+	<T> T set(AttachmentType<T> attachment, T value);
+
+	/**
+	 * Remove the value associated to the passed attachment type, and return it, or {@code null} if there was no previous value.
+	 */
+	@Nullable
+	<T> T remove(AttachmentType<T> attachment);
```

Comment:
Imo the default state of an attachment shouldn't be null, but should be regulated by the type.

Reply:
That's going to be too messy when the attachment wants to capture a reference to the entity imo. `computeIfAbsent` should be fine. If this proves to be too bad we can always go for a v2 API... this PR only adds 2 API classes, so not the end of the world.